### PR TITLE
Increase the S_max upper limit and handle case for N > 100

### LINKED ISSUE #13334: poisson_conf_interval 'kraft-burrows-nousek' fails unexpectedly
### Expected behavior

poisson_conf_interval 'kraft-burrows-nousek'  calculated values or fails with an obvious error message.

### Actual behavior

The error message comes form deep within `scipy.optimize` and it's not clear for a user how to react. It's particularly confusing here because, at least on my platform, the function could work if not for a hardcoded upper value of an interval. The problem is in line 1160 `S_max = brentq(func, N - B, 100)` (see pasted error output below). In the function, the upper end of the interval that `brentq` searches is hardcoded to 100. That's because (looking at the code comments and also speaking from memory as the person who originally implemented this) much larger value lead to numerical overflow error - scipy simple can't be used. However, for the numbers in my example, the calculation works when changing the hardcoded `100` to `200`.

### Notes for tackling this
So, this looks very simple - just increase the hardcoded `100`. However, we want to be a bit careful not to increase it too much because nobody is helped by running into overflow's instead. So, one might instead have to decrease that "Don't use scipy with N > 100" rule to a smaller number  - finding the balance here might require some experimentation and new numbers in the test.

### Steps to Reproduce

With scipy installed:

```python
from astropy.stats import poisson_conf_interval
poisson_conf_interval(n=int(98), background=3.8, confidence_level=0.95,
                                             interval='kraft-burrows-nousek')
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Input In [48], in <cell line: 1>()
----> 1 poisson_conf_interval(n=int(98), background=3.8, confidence_level=0.95,
      2                                              interval='kraft-burrows-nousek')

File /nfs/melkor/d1/guenther/soft/mambaforge/envs/ciao-4.14/lib/python3.9/site-packages/astropy/stats/funcs.py:764, in poisson_conf_interval(n, interval, sigma, background, confidence_level)
    762     if np.any(background < 0):
    763         raise ValueError('Background must be >= 0.')
--> 764     conf_interval = np.vectorize(_kraft_burrows_nousek,
    765                                  cache=True)(n, background, confidence_level)
    766     conf_interval = np.vstack(conf_interval)
    767 else:

File /nfs/melkor/d1/guenther/soft/mambaforge/envs/ciao-4.14/lib/python3.9/site-packages/numpy/lib/function_base.py:2304, in vectorize.__call__(self, *args, **kwargs)
   2301     vargs = [args[_i] for _i in inds]
   2302     vargs.extend([kwargs[_n] for _n in names])
-> 2304 return self._vectorize_call(func=func, args=vargs)

File /nfs/melkor/d1/guenther/soft/mambaforge/envs/ciao-4.14/lib/python3.9/site-packages/numpy/lib/function_base.py:2382, in vectorize._vectorize_call(self, func, args)
   2380     res = func()
   2381 else:
-> 2382     ufunc, otypes = self._get_ufunc_and_otypes(func=func, args=args)
   2384     # Convert args to object arrays first
   2385     inputs = [asanyarray(a, dtype=object) for a in args]

File /nfs/melkor/d1/guenther/soft/mambaforge/envs/ciao-4.14/lib/python3.9/site-packages/numpy/lib/function_base.py:2342, in vectorize._get_ufunc_and_otypes(self, func, args)
   2338     raise ValueError('cannot call `vectorize` on size 0 inputs '
   2339                      'unless `otypes` is set')
   2341 inputs = [arg.flat[0] for arg in args]
-> 2342 outputs = func(*inputs)
   2344 # Performance note: profiling indicates that -- for simple
   2345 # functions at least -- this wrapping can almost double the
   2346 # execution time.
   2347 # Hence we make it optional.
   2348 if self.cache:

File /nfs/melkor/d1/guenther/soft/mambaforge/envs/ciao-4.14/lib/python3.9/site-packages/astropy/stats/funcs.py:1290, in _kraft_burrows_nousek(N, B, CL)
   1288 if HAS_SCIPY and N <= 100:
   1289     try:
-> 1290         return _scipy_kraft_burrows_nousek(N, B, CL)
   1291     except OverflowError:
   1292         if not HAS_MPMATH:

File /nfs/melkor/d1/guenther/soft/mambaforge/envs/ciao-4.14/lib/python3.9/site-packages/astropy/stats/funcs.py:1160, in _scipy_kraft_burrows_nousek(N, B, CL)
   1157     out = eqn9_left(s_min, s, N, B)
   1158     return out[0] - CL
-> 1160 S_max = brentq(func, N - B, 100)
   1161 S_min = find_s_min(S_max, N, B)
   1162 return S_min, S_max

File /nfs/melkor/d1/guenther/soft/mambaforge/envs/ciao-4.14/lib/python3.9/site-packages/scipy/optimize/_zeros_py.py:783, in brentq(f, a, b, args, xtol, rtol, maxiter, full_output, disp)
    781 if rtol < _rtol:
    782     raise ValueError("rtol too small (%g < %g)" % (rtol, _rtol))
--> 783 r = _zeros._brentq(f, a, b, xtol, rtol, maxiter, args, full_output, disp)
    784 return results_c(full_output, r)

ValueError: f(a) and f(b) must have different signs
```

### System Details
Linux-3.10.0-1062.12.1.el7.x86_64-x86_64-with-glibc2.17
Python 3.9.7 (default, Sep 16 2021, 13:09:58) 
[GCC 7.5.0]
Numpy 1.21.2
pyerfa 2.0.0
astropy 5.1
Scipy 1.7.1

### PR DESCRIPTION
**Pull request by salmanhiro**
### Description

This pull request is to address `S_max` upper limit and handle case for `N > 100` in poisson_conf_interval with `kraft-burrows-nousek` interval

Fixes #13334

Description:
This pull request extends the range of the S_max calculation in the _scipy_kraft_burrows_nousek function to handle cases where N - B is more than 100. The current implementation uses the brentq optimization method, but this method has a limit on the upper value of the range.

Changes Made:
- Modified the upper limit of the brentq range to a user-defined max_value, allowing for larger values of N - B.
- Added a comment to explain the purpose of the change and the reason for choosing the new max_value.

Testing:
- Added test cases for the extended range for scipy and mpmath

Note for review:
- Is there a better way theoretically than to hardcode the `S_max`? 

- [x] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

**Comment by pllim** [comment]
Thanks! Needs a change log. Please see https://github.com/astropy/astropy/blob/main/docs/changes/README.rst

File: astropy/stats/funcs.py [hunk 1]
```diff
@@ -1247,12 +1247,20 @@ def _kraft_burrows_nousek(N, B, CL):
...
-            if not HAS_MPMATH:
-                raise ValueError("Need mpmath package for input numbers this large.")
+    if HAS_SCIPY:
+        if N <= 100:
+            try:
+                return _scipy_kraft_burrows_nousek(N, B, CL)
+            except OverflowError:
+                if not HAS_MPMATH:
+                    raise ValueError("Need mpmath package for input numbers this large.")
+        else:
```
Comment (hamogu) [inline astropy/stats/funcs.py:1257]: In this implementation, why is there is if..else block if the code that's executed is the same in the "if" as it is in the "else"? If that's what you want, no "if" is needed.

**Comment by hamogu** [comment]
Well, as the author of the original code and the author of #13334 you might think that I knew what `S_max` should be. I can tell you why I put it there: In the formulas that are evaluated in this code, there are a lot of terms that get really large; even though the final answer might not be that large; essentially it's dividing two large number. Unfortunately, there is no way (as far as I know, believe I tried) to analytically simplify the expression. All the large numbers need to be evaluated numerically to get to a result.

Once numbers get too big, Scipy will run into a numerical overflow, so we don't want that. (mpmath on the other hand can work in arbitrary precision, so there is no numerical limit like this until your computer runs out of memory). Now, which inputs are so big that we get to the numerical overflow? I just tried out a few numbers, and found a limit of "S_max < 100" to work pretty well when I wrote the code. Unfortunately, as I found out in #13334 that's not the whole story. You can break it with smaller S_max, depending on the other input numbers, on the other hand, if the other parameters are favourable, S_max > 100 might work.

So, either we need to understand the formulas better, or use an approach like you try in this PR - don't have an S_max limit and just fail when an overflow happens. I don't know why I did not do that originally. 

I think what needs to be done here is to run numbers with mpmath and scipy and see that we always get the same answer when scipy gives and answer (and not an overflow error). 

If that's true, then this approach in this PR could be a good solution.

File: astropy/stats/funcs.py (hunk 1 above)
Comment (salmanhiro) [reply astropy/stats/funcs.py:1257]: Okay, at first I want to handle the not overflow but N > 100 case but now I think that's double.

**Comment by ohyos-isolt** [comment]
The error given in #13334 is caused by a ValueError, rather than a OverflowError. Thus the implication doesn't seem to solve the issue.

**Pull request closed without merging**